gusb/gusb-device.c

```diff
@@ -1807,9 +2101,58 @@ g_usb_device_bulk_transfer_async(GUsbDevice *self,
 	GcmDeviceReq *req;
 	gint rc;
 	GError *error = NULL;
+	GUsbDeviceEvent *event = NULL;
+	g_autofree gchar *event_id = NULL;
 
 	g_return_if_fail(G_USB_IS_DEVICE(self));
 
+	/* build event key either for load or save */
+	if (priv->device == NULL ||
+	    g_usb_context_get_flags(priv->context) & G_USB_CONTEXT_FLAGS_SAVE_EVENTS) {
+		g_autofree gchar *data_base64 = g_base64_encode(data, length);
+		event_id = g_strdup_printf("BulkTransfer:"
+					   "Endpoint=0x%02x,"
+					   "Data=%s,"
+					   "Length=0x%x",
+					   endpoint,
+					   data_base64,
+					   (guint)length);
+	}
+
+	/* emulated */
+	if (priv->device == NULL) {
+		GBytes *bytes;
+		event = g_usb_device_load_event(self, event_id);
+		if (event == NULL) {
+			g_task_report_new_error(self,
+						callback,
+						user_data,
+						g_usb_device_control_transfer_async,
+						G_IO_ERROR,
+						G_IO_ERROR_INVALID_DATA,
+						"no matching event for %s",
+						event_id);
+			return;
+		}
+		bytes = g_usb_device_event_get_bytes(event);
+		if (bytes == NULL) {
+			g_task_report_new_error(self,
+						callback,
+						user_data,
+						g_usb_device_control_transfer_async,
+						G_IO_ERROR,
+						G_IO_ERROR_INVALID_DATA,
+						"no matching event data for %s",
+						event_id);
+			return;
+		}
+		task = g_task_new(self, cancellable, callback, user_data);
+		memcpy(data, g_bytes_get_data(bytes, NULL), g_bytes_get_size(bytes));
```

Comment:
I don't think that a memcpy here is correct, I am actually surprised your code doesn't crash because it is sending out a static buffer using a bulk transfer. And then here you copy the stored data into its read-only memory.

i.e. if it is an OUT endpoint, do a memcmp and check that the data is what you are expecting. If it is an IN endpoint do the memcpy as you are doing ithere. That said, you *might* want to check that the passed buffer is actually large enough.

Reply:
> And then here you copy the stored data into its read-only memory

Oh? I'm copying the loaded GBytes into the mutable `data` that's passed into the `g_usb_device_control_transfer_async()` -- I think it's safe as the length of the buffer is stored in the event key too -- i.e. the buffer is known big enough. I guess an extra assert wouldn't hurt...

Reply:
But an outgoing transfer is known to be never read. In libfprint, the code will very often start an outgoing bulk transfer from a static buffer. So something like (except that libfprint has a thin libgusb wrapper):

```
static const char capture_msg[] = { 0x01, 0x02, 0x03 };
...
g_usb_device_bulk_transfer_async (device, EP_OUT, capture_msg, sizeof(capture_msg), 0, NULL, cb, self);
```

Reply:
But g_usb_device_bulk_transfer_async() needs a mutable `guint8 *` -- how does the compiler not get upset when you pass in a `const char[]`?

Reply:
> you need to do a cast to drop the const

Dude, **there be dragons**!

As a workaround for this borderline pathological (in the funniest sense) behaviour, how about we avoid the memcpy is skipped if the buffer exactly matches the event data?